### ISSUE TITLE
Problems with v2.0.0

### ISSUE THREAD

**Issue by cvn** [TTLabs/EvaporateJS#276]
Since 2.0.0-rc.7, I have an issue where after cancelling uploads, uploading breaks.

1. Start two 100MB uploads
   - upload1 starts uploading, upload2 is queued
2. cancel upload1, then upload2 rapidly
    - I delete stubs on my server, in parallel, then cancel. <100ms between events is usual
    - This logs twice: `Requeued my-bucket/upload2.jpg status: 20 reason: file done`
3. Start upload3
    - This logs: `Requeued my-bucket/upload2.jpg status: 20 reason: first file`
    - upload3 does not start

This is a follow-up to #267.

**Comment by bikeath1337** [comment]
Does reproducing this require that the cancel event has to happen in rapid succession? If so, I'm not sure how to test the use case.

And what happened with the Network logs? Did the 2 files cancel? And is there any sign of activity for upload3?

**Comment by cvn** [comment]
Sorry for the sparse description.

This code, with file sizes > partSize * maxParts, will trigger the problem:
```js
evap.add({name: 'upload1.zip' ... });
evap.add({name: 'upload2.zip' ... });
setTimeout(function () { evap.cancel('my-bucket/upload1.zip'); }, 5000);
setTimeout(function () { evap.cancel('my-bucket/upload2.zip'); }, 5100);
setTimeout(function () { evap.add({name: 'upload3.zip' ... }); }, 10000);
```
This is the log:
```
my-app.js:206 cancel my-bucket/upload1.zip @ 1480649392151

evaporate.js:595 stopping FileUpload  my-bucket/upload1.zip
4 x evaporate.js:1092   ### 6291456
evaporate.js:1092   ### 5198329
evaporate.js:1586 V2 stringToSign: DELETE
x-amz-date:Fri, 02 Dec 2016 03:29:52 GMT
/my-bucket/upload1.zip?uploadId=...
evaporate.js:1152 abort signature: q7cyE2YocOxp2xYVKHfkD6io1Q8=

my-app.js:206 cancel my-bucket/upload2.zip @ 1480649392256

evaporate.js:595 stopping FileUpload  my-bucket/upload2.zip
evaporate.js:187 Requeued upload2.zip status: 20 reason: file done
evaporate.js:270 Uncaught (in promise) User aborted the upload(anonymous function) @ evaporate.js:270
evaporate.js:187 Requeued upload2.zip status: 20 reason: file done
evaporate.js:270 Uncaught (in promise) User aborted the upload(anonymous function) @ evaporate.js:270

my-app.js:166 add my-bucket/upload3.zip @ 1480649397151

evaporate.js:187 Requeued upload2.zip status: 20 reason: first file
```

upload1 and upload2 do cancel successfully. the DELETE for upload1 returns 204, there is no network activity at all for upload2 or upload3.

**Comment by bikeath1337** [comment]
Very helpful. Thank you.

**Comment by bikeath1337** [comment]
Thank you for the clear instructions, @cvn . I identified the root cause and addressed it in #277. Please feel free to try it out.

**Comment by cvn** [comment]
That seems to fix the previous case, but here's a new problem case. With this code, upload3 does not cancel. It keeps uploading till completion. (Actually, once in 10 runs, all 3 uploads cancelled successfully, but I suspect you will be able to reproduce the issue easily)

```js
evap.add({name: 'upload1.zip' ... });
evap.add({name: 'upload2.zip' ... });
evap.add({name: 'upload3.zip' ... });
setTimeout(function () { evap.cancel('my-bucket/upload1.zip'); }, 5000);
setTimeout(function () { evap.cancel('my-bucket/upload2.zip'); }, 5100);
setTimeout(function () { evap.cancel('my-bucket/upload3.zip'); }, 5200);
```
Here is the log:
```
my-app.js:207 cancel my-bucket/upload1.zip @ 1480749505180

evaporate.js:596 stopping FileUpload  my-bucket/upload1.zip
4evaporate.js:1093   ### 6291456
evaporate.js:1093   ### 5198329
evaporate.js:1587 V2 stringToSign: DELETE
x-amz-date:Sat, 03 Dec 2016 07:18:25 GMT
/my-bucket/upload1.zip?uploadId=...
evaporate.js:1153 abort signature: ET0QiDPf41N/loFPhyXqMVcwiQ0=

my-app.js:207 cancel my-bucket/upload2.zip @ 1480749505285

evaporate.js:596 stopping FileUpload  my-bucket/upload2.zip
evaporate.js:182 Starting upload3.zip reason: file done
evaporate.js:572 
evaporate.js:1587 V2 stringToSign: POST
application/zip
x-amz-date:Sat, 03 Dec 2016 07:18:25 GMT
/my-bucket/upload3.zip?uploads
evaporate.js:271 Uncaught (in promise) User aborted the upload(anonymous function) @ evaporate.js:271
evaporate.js:1153 initiate signature: ONTF6NigdJVFMiY0x3k04aTGXyE=

my-app.js:207 cancel my-bucket/upload3.zip @ 1480749505384

evaporate.js:596 stopping FileUpload  my-bucket/upload3.zip
evaporate.js:271 Uncaught (in promise) User aborted the upload(anonymous function) @ evaporate.js:271
evaporate.js:271 Uncaught (in promise) User aborted the upload(anonymous function) @ evaporate.js:271
evaporate.js:1234 InitiateMultipartUpload ID is ...
evaporate.js:1494 1 "payload bytes:" 0 "->" 6291456 ", size:" 6291456
evaporate.js:1494 2 "payload bytes:" 6291456 "->" 12582912 ", size:" 6291456
evaporate.js:1494 3 "payload bytes:" 12582912 "->" 18874368 ", size:" 6291456
evaporate.js:1494 4 "payload bytes:" 18874368 "->" 25165824 ", size:" 6291456
evaporate.js:1494 5 "payload bytes:" 25165824 "->" 31457280 ", size:" 4375064
evaporate.js:1397 uploadPart # 1 submitting
evaporate.js:1382 upload #1 MD5 digest: fHVSNBLYYiRZC6a1gvcgpQ==
evaporate.js:1406 Sending upload #1
evaporate.js:1587 V2 stringToSign: PUT
fHVSNBLYYiRZC6a1gvcgpQ==
x-amz-date:Sat, 03 Dec 2016 07:18:25 GMT
/my-bucket/upload3.zip?partNumber=1&uploadId=...
evaporate.js:1153 upload #1 signature: TbVX833nADy2hQ8euASfdzuDiQo=
evaporate.js:1494 1 "payload bytes:" 0 "->" 6291456 ", size:" 6291456
evaporate.js:1397 uploadPart # 2 submitting

... more upload logging for /my-bucket/upload3.zip here ...

evaporate.js:1153 complete signature: 6KBBd3mRafU/wZPde8oXETksJH8=
evaporate.js:1093   ### undefined
```

**Comment by bikeath1337** [comment]
Thanks for the helpful test cases. I pushed a new branch that resolves them and further improves the code base. We needed to check the status of the file more frequently because the order of execution is indeterminate due to the asynchronous nature of Evaporate.

There were two issues: several methods were continuing even though the file was aborted and the record of the upload in localStorage was being written sometimes even though the upload was aborted.

I updated #277 with the fixes.

**Comment by cvn** [comment]
So far so good with cancelling in the latest update. A new issue is that progress becomes inaccurate after force pause.

```js
evap.add({
    name: 'upload1.zip',
    progress: function (p) {console.log('p:', p);},
    ...
});
setTimeout(function () { evap.pause('my-bucket/upload1.zip', {force:true}); }, 5000);
setTimeout(function () { evap.resume('my-bucket/upload1.zip'); }, 6000);
```

```
...
my-app.js:178 p: 0.1408313283100635
my-app.js:178 p: 0.1645730081784267

my-app.js:197 pause my-bucket/upload1.zip @ 1480798336595

evaporate.js:611 pausing FileUpload, force: true my-bucket/upload1.zip
evaporate.js:1097   ### 6291456
my-app.js:178 p: 0.1705084281455175
...

my-app.js:202 resume my-bucket/upload1.zip @ 1480798339388

evaporate.js:182 Starting rxjp3zr3ewhd/upload1.zip reason: first file
my-app.js:178 p: 0.1705084281455175
evaporate.js:539 resuming FileUpload  my-bucket/upload1.zip
evaporate.js:1401 uploadPart # 1 submitting
...
my-app.js:178 p: 0.3534272798585885
my-app.js:178 p: 0.3760897924602079
...
my-app.js:178 p: 1.1327375738094851
my-app.js:178 p: 1.1543209191443609
my-app.js:178 p: 1.1705084281455176
```

**Comment by bikeath1337** [comment]
Thank you. I think the commit I just pushed addresses the issue., @cvn.

**Comment by cvn** [comment]
The latest version is solid. All my previous issues are resolved and I have found no new ones.

Great work @bikeath1337

**Comment by bikeath1337** [comment]
Much obliged, @cvn.
